flaky: enterprise/replicasync race detected during execution of test

Spotted in: https://github.com/coder/coder/actions/runs/3605282480/jobs/6075510642

```
=== Failed
=== FAIL: enterprise/replicasync TestReplica/TwentyConcurrent (1.27s)
    testing.go:1319: race detected during execution of test
    --- FAIL: TestReplica/TwentyConcurrent (1.27s)

=== FAIL: enterprise/replicasync TestReplica (0.00s)

DONE 2515 tests, 11 skipped, 2 failures in 162.177s
make: *** [Makefile:476: test-postgres] Error 1
Error: Process completed with exit code 2.
```

--- a/enterprise/replicasync/replicasync.go
+++ b/enterprise/replicasync/replicasync.go
@@ -276,6 +276,8 @@ func (m *Manager) syncReplicas(ctx context.Context) error {
 		return xerrors.Errorf("ping database: %w", err)
 	}
 
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
 	replica, err := m.db.UpdateReplica(ctx, database.UpdateReplicaParams{
 		ID:              m.self.ID,
 		UpdatedAt:       database.Now(),
@@ -291,8 +293,6 @@ func (m *Manager) syncReplicas(ctx context.Context) error {
 	if err != nil {
 		return xerrors.Errorf("update replica: %w", err)
 	}
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
 	if m.self.Error != replica.Error {
 		// Publish an update occurred!
 		err = m.pubsub.Publish(PubsubEvent, []byte(m.self.ID.String()))

--- a/enterprise/replicasync/replicasync_test.go
+++ b/enterprise/replicasync/replicasync_test.go
@@ -206,7 +206,12 @@ func TestReplica(t *testing.T) {
 				_ = server.Close()
 			})
 			done := false
+
+			var m sync.Mutex
 			server.SetCallback(func() {
+				m.Lock()
+				defer m.Unlock()
+
 				if len(server.All()) != count {
 					return
 				}

